File: app/src/main/java/javinator9889/bitcoinpools/FragmentViews/Tab1PoolsChart.java
package javinator9889.bitcoinpools.FragmentViews;

import android.graphics.Color;
import android.graphics.Typeface;
import android.os.Bundle;
import android.os.Handler;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;
import android.widget.Toast;

import com.afollestad.materialdialogs.MaterialDialog;
import com.github.mikephil.charting.animation.Easing;
import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.components.Description;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;
import com.github.mikephil.charting.utils.ColorTemplate;
import com.google.firebase.perf.metrics.AddTrace;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import javinator9889.bitcoinpools.BitCoinApp;
import javinator9889.bitcoinpools.Constants;
import javinator9889.bitcoinpools.MainActivity;
import javinator9889.bitcoinpools.R;

import static javinator9889.bitcoinpools.Constants.SHARED_PREFERENCES.CUSTOM_POOLS;

/**
 * Created by Javinator9889 on 28/01/2018. Creates view for main chart (pools chart)
 */

public class Tab1PoolsChart extends BaseFragment {
    private static Map<String, Float> RETRIEVED_DATA = new LinkedHashMap<>();
    private static ViewGroup.LayoutParams TABLE_PARAMS;
    private static float MARKET_PRICE_USD;
    private PieChart destinationChart;
    private Thread tableThread;
    private int mMaximumPoolsToShow;
    private TextView mPoolsText;
    private MaterialDialog md;

    public Tab1PoolsChart() {
    }

    @SuppressWarnings("unchecked")
    public static Tab1PoolsChart newInstance(Object... params) {
        Bundle args = new Bundle();
        args.putFloat("MPU", Float.valueOf(String.valueOf(params[0])));
        args.putSerializable("RD", (HashMap<String, Float>) params[1]);
        Tab1PoolsChart fragment = new Tab1PoolsChart();
        fragment.setArguments(args);
        return fragment;
    }

    @SuppressWarnings("unchecked")
    @Override
    @AddTrace(name = "onCreateViewForTab1")
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        final View createdView = inflater.inflate(R.layout.poolschart, container, false);

        destinationChart = createdView.findViewById(R.id.chart);
        mPoolsText = createdView.findViewById(R.id.showingNumber);

        final TableLayout tableLayout = createdView.findViewById(R.id.poolstable);

        mMaximumPoolsToShow = BitCoinApp.getSharedPreferences().getInt(CUSTOM_POOLS, 10);
        MARKET_PRICE_USD = getArguments().getFloat("MPU");
        RETRIEVED_DATA = (HashMap<String, Float>) getArguments().getSerializable("RD");

        initT(createdView);
        createPieChart();
        createTable(tableLayout, createdView);
        try {
            tableThread.join();
            return createdView;
        } catch (InterruptedException e) {
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        DisplayMetrics metrics = getResources().getDisplayMetrics();
        int dpHeight = metrics.heightPixels;
        int finalDp = (int) (dpHeight * 0.5);
        ConstraintLayout.LayoutParams params = (ConstraintLayout.LayoutParams) destinationChart
                .getLayoutParams();
        params.height = finalDp;
        params.matchConstraintMaxHeight = dpHeight;
        destinationChart.setLayoutParams(params);
        destinationChart.invalidate();

        View duplicatedView = View.inflate(view.getContext(), R.layout.poolschart, null);
        final TableLayout tableLayout = duplicatedView.findViewById(R.id.poolstable);
        createTable(tableLayout, duplicatedView);
        try {
            tableThread.join();
            md = new MaterialDialog.Builder(view.getContext())
                    .title(R.string.latest_24h_pools_information)
                    .customView(tableLayout, true)
                    .positiveText(R.string.accept)
                    .positiveColor(Color.BLACK)
                    .cancelable(true)
                    .build();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        ConstraintLayout container = view.findViewById(R.id.constraintLayout);
        container.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Toast.makeText(view.getContext(), R.string.show_longer_table, Toast
                        .LENGTH_LONG).show();
            }
        });
        container.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                md.show();
                return false;
            }
        });
//        md.show();
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (isAdded() && isVisibleToUser) {
            MainActivity.MAINACTIVITY_TOOLBAR.setTitle(getString(R.string.BTCP) + MARKET_PRICE_USD);
        }
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (isAdded() && isVisible()) {
            MainActivity.MAINACTIVITY_TOOLBAR.setTitle(getString(R.string.BTCP) + MARKET_PRICE_USD);
        }
    }

    private void createPieChart() {
        new Handler().postDelayed(new Runnable() {  // Required for animation
            @Override
            public void run() {
                Log.d(Constants.LOG.MATAG, Constants.LOG.LOADING_CHART);
                List<PieEntry> values = new ArrayList<>();
                List<Map.Entry<String, Float>> entryList = new ArrayList<>(
                        RETRIEVED_DATA.entrySet());
                Map.Entry<String, Float> getEntry;
                int count = 0;
                for (int i = entryList.size() - 1; (i >= 0) && (count < mMaximumPoolsToShow); --i) {
                    getEntry = entryList.get(i);
                    Log.i(Constants.LOG.MATAG, "Accessing at: " + i + " | Key: "
                            + getEntry.getKey() + " | Value: " + getEntry.getValue());
                    values.add(new PieEntry(getEntry.getValue(), getEntry.getKey()));
                    ++count;
                }
                PieDataSet data = new PieDataSet(values, "Latest 24h BTC pools");
                data.setColors(ColorTemplate.MATERIAL_COLORS);
                data.setValueTextSize(10f);

                try {
                    destinationChart.setData(new PieData(data));
                    destinationChart.setUsePercentValues(true);
                    destinationChart.setEntryLabelColor(ColorTemplate.rgb("#000000"));
                    Description description = new Description();
                    description.setText(getString(R.string.porcent));
                    destinationChart.setDescription(description);
                    destinationChart.getLegend().setEnabled(false);
                    destinationChart.setDragDecelerationFrictionCoef(0.95f);
                    destinationChart.setHardwareAccelerationEnabled(true);
                    destinationChart.setMinimumWidth(10);
                    destinationChart.animateY(1400, Easing.EaseInOutQuad);
                } catch (Exception e) {
                    Log.e("PieChart", "Error loading PieChart: " + e.getMessage());
                }
            }
        }, 100);
    }

    private void createTable(final TableLayout destinationTable, final View view) {
        tableThread = new Thread() {
            public void run() {
                Log.d(Constants.LOG.MATAG, Constants.LOG.LOADING_TABLE);
                List<Map.Entry<String, Float>> entryList = new ArrayList<>(
                        RETRIEVED_DATA.entrySet());
                Map.Entry<String, Float> getEntry;
                int count = 1;

                TableRow fetchTableRow = view.findViewById(R.id.masterRow);
                TextView firstPool = new TextView(view.getContext());
                TextView firstBlock = new TextView(view.getContext());

                getEntry = entryList.get(entryList.size() - 1);

                firstPool.setText(getEntry.getKey());
                firstPool.setTypeface(Typeface.DEFAULT_BOLD);
                firstPool.setTextSize(16f);

                firstBlock.setText(String.valueOf(getEntry.getValue()));
                firstBlock.setTypeface(Typeface.MONOSPACE);
                firstBlock.setTextSize(16f);

                fetchTableRow.addView(firstPool);
                fetchTableRow.addView(firstBlock);
                for (int i = entryList.size() - 2; (i >= 0) && (count < mMaximumPoolsToShow); --i) {
                    TextView poolName = new TextView(view.getContext());
                    TextView poolBlock = new TextView(view.getContext());
                    TableRow tableRow = new TableRow(view.getContext());

                    getEntry = entryList.get(i);
                    poolName.setText(getEntry.getKey());
                    poolBlock.setText(String.valueOf(getEntry.getValue()));

                    poolName.setTypeface(Typeface.DEFAULT_BOLD);
                    poolName.setTextSize(16f);

                    poolBlock.setTypeface(Typeface.MONOSPACE);
                    poolBlock.setTextSize(16f);

                    tableRow.setLayoutParams(TABLE_PARAMS);
                    tableRow.addView(poolName);
                    tableRow.addView(poolBlock);

                    destinationTable.addView(tableRow);
                    ++count;
                }
                destinationTable.invalidate();
                mPoolsText.setText(getString(R.string.number_of_pools_displayed,
                        count,
                        entryList.size()));
            }
        };
        tableThread.setName("table_thread");
        tableThread.start();
    }

    private void initT(View view) {
        TableRow masterRow = view.findViewById(R.id.masterRow);
        TABLE_PARAMS = masterRow.getLayoutParams();
    }
}

File: app/src/main/java/javinator9889/bitcoinpools/FragmentViews/Tab1PoolsChart.kt
package javinator9889.bitcoinpools.FragmentViews

import android.graphics.Color
import android.graphics.Typeface
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.TextView
import android.widget.Toast
import androidx.constraintlayout.widget.ConstraintLayout
import com.afollestad.materialdialogs.MaterialDialog
import com.github.mikephil.charting.animation.Easing
import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.components.Description
import com.github.mikephil.charting.data.PieData
import com.github.mikephil.charting.data.PieDataSet
import com.github.mikephil.charting.data.PieEntry
import com.github.mikephil.charting.utils.ColorTemplate
import com.google.firebase.perf.metrics.AddTrace
import javinator9889.bitcoinpools.BitCoinApp
import javinator9889.bitcoinpools.Constants
import javinator9889.bitcoinpools.Constants.SHARED_PREFERENCES.CUSTOM_POOLS
import javinator9889.bitcoinpools.MainActivity
import javinator9889.bitcoinpools.R
import java.util.*

/**
 * Created by Javinator9889 on 28/01/2018. Creates view for main chart (pools chart)
 */

class Tab1PoolsChart : BaseFragment() {
    private var destinationChart: PieChart? = null
    private var tableThread: Thread? = null
    private var mMaximumPoolsToShow: Int = 0
    private var mPoolsText: TextView? = null
    private var md: MaterialDialog? = null

    @AddTrace(name = "onCreateViewForTab1")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val createdView = inflater.inflate(R.layout.poolschart, container, false)

        destinationChart = createdView.findViewById(R.id.chart)
        mPoolsText = createdView.findViewById(R.id.showingNumber)

        val tableLayout = createdView.findViewById<TableLayout>(R.id.poolstable)

        mMaximumPoolsToShow = BitCoinApp.getSharedPreferences().getInt(CUSTOM_POOLS, 10)
        MARKET_PRICE_USD = arguments!!.getFloat("MPU")
        RETRIEVED_DATA = arguments!!.getSerializable("RD") as HashMap<String, Float>

        initT(createdView)
        createPieChart()
        createTable(tableLayout, createdView)
        try {
            tableThread!!.join()
            return createdView
        } catch (e: InterruptedException) {
            e.printStackTrace()
            return null
        }

    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val metrics = resources.displayMetrics
        val dpHeight = metrics.heightPixels
        val finalDp = (dpHeight * 0.5).toInt()
        val params = destinationChart!!
                .layoutParams as ConstraintLayout.LayoutParams
        params.height = finalDp
        params.matchConstraintMaxHeight = dpHeight
        destinationChart!!.layoutParams = params
        destinationChart!!.invalidate()

        val duplicatedView = View.inflate(view.context, R.layout.poolschart, null)
        val tableLayout = duplicatedView.findViewById<TableLayout>(R.id.poolstable)
        createTable(tableLayout, duplicatedView)
        try {
            tableThread!!.join()
            md = MaterialDialog.Builder(view.context)
                    .title(R.string.latest_24h_pools_information)
                    .customView(tableLayout, true)
                    .positiveText(R.string.accept)
                    .positiveColor(Color.BLACK)
                    .cancelable(true)
                    .build()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

        val container = view.findViewById<ConstraintLayout>(R.id.constraintLayout)
        container.setOnClickListener {
            Toast.makeText(view.context, R.string.show_longer_table, Toast
                    .LENGTH_LONG).show()
        }
        container.setOnLongClickListener {
            md!!.show()
            false
        }
        //        md.show();
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (isAdded && isVisibleToUser) {
            MainActivity.MAINACTIVITY_TOOLBAR.title = getString(R.string.BTCP) + MARKET_PRICE_USD
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (isAdded && isVisible) {
            MainActivity.MAINACTIVITY_TOOLBAR.title = getString(R.string.BTCP) + MARKET_PRICE_USD
        }
    }

    private fun createPieChart() {
        Handler().postDelayed(// Required for animation
                {
                    Log.d(Constants.LOG.MATAG, Constants.LOG.LOADING_CHART)
                    val values = ArrayList<PieEntry>()
                    val entryList = ArrayList<Entry<String, Float>>(
                            RETRIEVED_DATA.entries)
                    var getEntry: Entry<String, Float>
                    var count = 0
                    var i = entryList.size - 1
                    while (i >= 0 && count < mMaximumPoolsToShow) {
                        getEntry = entryList[i]
                        Log.i(Constants.LOG.MATAG, "Accessing at: " + i + " | Key: "
                                + getEntry.key + " | Value: " + getEntry.value)
                        values.add(PieEntry(getEntry.value, getEntry.key))
                        ++count
                        --i
                    }
                    val data = PieDataSet(values, "Latest 24h BTC pools")
                    data.setColors(*ColorTemplate.MATERIAL_COLORS)
                    data.valueTextSize = 10f

                    try {
                        destinationChart!!.data = PieData(data)
                        destinationChart!!.setUsePercentValues(true)
                        destinationChart!!.setEntryLabelColor(ColorTemplate.rgb("#000000"))
                        val description = Description()
                        description.text = getString(R.string.porcent)
                        destinationChart!!.description = description
                        destinationChart!!.legend.isEnabled = false
                        destinationChart!!.dragDecelerationFrictionCoef = 0.95f
                        destinationChart!!.setHardwareAccelerationEnabled(true)
                        destinationChart!!.minimumWidth = 10
                        destinationChart!!.animateY(1400, Easing.EaseInOutQuad)
                    } catch (e: Exception) {
                        Log.e("PieChart", "Error loading PieChart: " + e.message)
                    }
                }, 100)
    }

    private fun createTable(destinationTable: TableLayout, view: View) {
        tableThread = object : Thread() {
            override fun run() {
                Log.d(Constants.LOG.MATAG, Constants.LOG.LOADING_TABLE)
                val entryList = ArrayList<Entry<String, Float>>(
                        RETRIEVED_DATA.entries)
                var getEntry: Entry<String, Float>
                var count = 1

                val fetchTableRow = view.findViewById<TableRow>(R.id.masterRow)
                val firstPool = TextView(view.context)
                val firstBlock = TextView(view.context)

                getEntry = entryList[entryList.size - 1]

                firstPool.setText(getEntry.key)
                firstPool.typeface = Typeface.DEFAULT_BOLD
                firstPool.textSize = 16f

                firstBlock.setText(getEntry.value.toString())
                firstBlock.typeface = Typeface.MONOSPACE
                firstBlock.textSize = 16f

                fetchTableRow.addView(firstPool)
                fetchTableRow.addView(firstBlock)
                var i = entryList.size - 2
                while (i >= 0 && count < mMaximumPoolsToShow) {
                    val poolName = TextView(view.context)
                    val poolBlock = TextView(view.context)
                    val tableRow = TableRow(view.context)

                    getEntry = entryList[i]
                    poolName.setText(getEntry.key)
                    poolBlock.setText(getEntry.value.toString())

                    poolName.typeface = Typeface.DEFAULT_BOLD
                    poolName.textSize = 16f

                    poolBlock.typeface = Typeface.MONOSPACE
                    poolBlock.textSize = 16f

                    tableRow.layoutParams = TABLE_PARAMS
                    tableRow.addView(poolName)
                    tableRow.addView(poolBlock)

                    destinationTable.addView(tableRow)
                    ++count
                    --i
                }
                destinationTable.invalidate()
                mPoolsText!!.text = getString(R.string.number_of_pools_displayed,
                        count,
                        entryList.size)
            }
        }
        tableThread!!.name = "table_thread"
        tableThread!!.start()
    }

    private fun initT(view: View) {
        val masterRow = view.findViewById<TableRow>(R.id.masterRow)
        TABLE_PARAMS = masterRow.layoutParams
    }

    companion object {
        private var RETRIEVED_DATA: Map<String, Float> = LinkedHashMap()
        private var TABLE_PARAMS: ViewGroup.LayoutParams? = null
        private var MARKET_PRICE_USD: Float = 0.toFloat()

        fun newInstance(vararg params: Any): Tab1PoolsChart {
            val args = Bundle()
            args.putFloat("MPU", java.lang.Float.valueOf(params[0].toString()))
            args.putSerializable("RD", params[1] as HashMap<String, Float>)
            val fragment = Tab1PoolsChart()
            fragment.arguments = args
            return fragment
        }
    }
}
